README
- Wallpapoz README -
#=============================================================================
#
#    Readme 
#    Copyright (C) 2011 Vajrasky Akbar Kok <akbarhome@gmail.com>
#
#=============================================================================
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
#=============================================================================

About
--------------------------------------------------------------------------
Wallpapoz application enables you to configure Gnome and XFCE desktop wallpapers
in unique way. You could have Gnome and XFCE desktop wallpaper changes when 
the specified time has passed. The most important feature is you could 
have Gnome and XFCE desktop wallpaper changes when you change workspace. It means 
you could group your wallpapers into specific workspace. It offers
quick orientation cues where you are. You can manage it so when you change 
to first workspace, your desktop wallpaper will be picked from wallpapers 
group for that workspace. When you change to second workspace, it will be
picked from wallpapers group for second workspace. So you could have 
situation like this. You change to third workspace and you will get cartoon
desktop wallpaper (wallpapers group for that workspace consists of cartoon
wallpapers only). You change to fourth workspace and you will get abstract
desktop wallpaper (wallpapers group for that workspace consists of abstract
wallpapers only). I think you get the idea. Off course you can disable this 
feature if you just want your desktop wallpaper changes when the specified 
time (specified by you) has passed and does not change when you change 
workspace.

Wallpapoz application consist of two programs. First is "wallpapoz". 
"wallpapoz" is the program to make the configuration file. What wallpapers 
do you want to use? It just like that. So before doing anything else, run
this program first. The file configuration is located in $HOME/.wallpapoz/
.wallpapoz.xml. Although you can edit the file with text editor, I don't
recommend it. If the file is corrupted, then remove it first before run
"wallpapoz" again. Don't run "wallpapoz" if the file is corrupted.

"daemon_wallpapoz" is the "real" program it self. The program will read the
configuration file and will change the wallpaper if you change the work-
space or the specified time has passed (depended on your options in
configuration file). 

You can run the "daemon_wallpapoz" from "wallpapoz" program. 
You can terminate it too from there.

If you are happy enough, you can start automatically running this program 
when you load the gnome desktop by adding "daemon_wallpapoz" to gnome session 
start up program. 

Dependencies
-----------------------------------------------------------------------------
If you run modern distro (released in 2005 year and above) with default Gnome 
desktop installed, your dependencies should already be fulfilled. It needs 
PyGtk, Python Glade, PIL (Python Imaging Library), and optionally Python Gnome to function.

Installation
-----------------------------------------------------------------------------
Run "python setup.py install" as root to install Wallpapoz in /usr/local.
To uninstall, simply run "python setup.py uninstall".

If you wish to install Wallpapoz in another base directory, use the --installdir
option. Example: "python setup.py install --installdir /usr".
If you have installed with --installdir, do not forget to use the same option
when uninstalling.

Example installation:

    $ tar -xzf wallpapoz-0.6.tar.gz
    $ cd wallpapoz-0.6

Depending on distro, the next steps could be this:
    # su
    # python setup.py install
or this:
    $ sudo python setup.py install

You don't have to install Wallpapoz to run the application. 
You can run Wallpapoz from its directory anywhere.

    $ cd wallpapoz-0.6
    $ cd src
    $ ./wallpapoz

You can put wallpapoz-0.6 directory in your home directory or anywhere.
Someday if you want to get rid of Wallpapoz, just remove the directory.

Credits and Thanks
-----------------------------------------------------------------------------
Translators: 
My self (Indonesian), Mamoru Tasaka (Japanese), Zoltán Nagy (Hungarian),
Nils Heymann (Germany), Daniel Holm (Swedish), Daniel Nylander (Swedish),
Yuri Myasoedov (Russian), Ivan Virgili (Italian), Lukáš Macháček (Czech),
Corentin Lemée (French), and Rene Garcia (Spanish and Portuguese).

The graphic icon is created by Stu Murray <suchawato@gmail.com>.

Ideas for improving professionalism of Wallpapoz project: 
Mamoru Tasaka

Leo von Klenze for fixing randomization bug.

And many more who report bugs, ask features, give supports, etc. Thank you.

Binary files license
-----------------------------------------------------------------------------
share/wallpapoz/glade/wallpapoz.png has the same license as the wallpapoz
package.
share/wallpapoz/wallpapoz.desktop has the same license as the wallpapoz
package.

Contact
-----------------------------------------------------------------------------
Vajrasky Akbar Kok <akbarhome@gmail.com>
http://vajrasky.wordpress.com/wallpapoz

Don't hesitate to contact me with any questions, suggestions, and bugs reports.


CHANGELOG
Version 0.6, 2011-09-18
-----------------------
- Changed homepage of wallpapoz.
- Added support for XFCE desktop and Gnome 3 desktop.
- Removed support for compiz.

Version 0.5, 2010-10-4
----------------------
- Added Hungarian, Germany, Russian, Italian, Czech,
Spanish and Portuguese translation.
- Added official icon.
- Randomization bug is fixed.

Version 0.4.1, 2007-6-4
-----------------------
- Swedish translation is updated by Daniel Nylander.
- French translation is added by Corentin Lemée.

Version 0.4, 2007-5-31
----------------------
- "First" release.


share/wallpapoz/lib/xml_processing.py
#================================================
#
#    xml_processing.py - Wallpapoz
#    Copyright (C) 2007 Vajrasky Akbar Kok <akbarhome@gmail.com>
#
#================================================
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
#================================================

## xml_processing.py -- manipulates xml file for gui or daemon

from xml.dom import minidom
import xml
import os
import sys
import gettext
import gtk
from wallpapoz_system import WallpapozSystem

# i18n
APP = "wallpapoz"
DIR = "../../locale"
gettext.bindtextdomain(APP, DIR)
gettext.textdomain(APP)
gtk.glade.bindtextdomain(APP, DIR)
gtk.glade.textdomain(APP)
_ = gettext.gettext

## XMLProcessing -- class for processing wallpapoz xml file
class XMLProcessing:

  ## The constructor
  def __init__(self):
    # our document instance
    self.xmldoc = None

    # the configuration file
    home = os.environ['HOME']
    self.config_file = home + "/.wallpapoz/wallpapoz.xml"

    # how many workspace do we have
    wallpapoz_system = WallpapozSystem()
    self.workspace_num = wallpapoz_system.get_total_workspaces()

    # our current wallpaper
    self.current_wallpaper = wallpapoz_system.finding_current_wallpaper()
    if self.current_wallpaper == '':
      self.current_wallpaper = _("change this with picture file")

    # if wallpapoz run for the first time ( no configuration file )
    # we make default list, for every workspace, we give one wallpaper that is our current wallpaper
    if not os.path.exists(self.config_file):
      print _("No configuration file. Use default configuration.")
      home = os.environ['HOME']
      if not os.path.exists(home + '/.wallpapoz'):
        os.makedirs(home + '/.wallpapoz')

      self.create_configuration_file(self.default_fill_list("treestore"), "treestore")

    # okay, parse the xml file
    try:
      self.xmldoc = minidom.parse(self.config_file)
    except xml.parsers.expat.ExpatError:
      print _("The configuration file is corrupted. Remove it then create a new one again with Wallpapoz!")
      sys.exit()

    # wallpapoz node
    self.wallpapoz_node = self.xmldoc.childNodes[1]

    # wallpapoz type ( workspace or desktop )
    try:
      self.wallpapoz_type = self.wallpapoz_node.attributes["type"].value
    except KeyError:
      self.wallpapoz_type = "workspace"

    # workspace node list
    # fill with workspace node if the type is workspace
    if self.wallpapoz_type == "workspace":
      self.workspace_node_list = self.wallpapoz_node.getElementsByTagName("workspace")
    # fill with file node if the type is desktop
    else:
      self.workspace_node_list = self.wallpapoz_node.getElementsByTagName("file")

  # reparse wallpapoz xml configuration file
  def reparse_xml(self):
    # okay, parse the xml file
    try:
      self.xmldoc = minidom.parse(self.config_file)
    except xml.parsers.expat.ExpatError:
      print _("The configuration file is corrupted. Remove it then create a new one again with Wallpapoz!")
      sys.exit()

    # wallpapoz node
    self.wallpapoz_node = self.xmldoc.childNodes[1]

    # wallpapoz type ( workspace or desktop )
    self.wallpapoz_type = self.wallpapoz_node.attributes["type"].value

    # workspace node list
    # fill with workspace node if the type is workspace
    if self.wallpapoz_type == "workspace":
      self.workspace_node_list = self.wallpapoz_node.getElementsByTagName("workspace")
    # fill with file node if the type is desktop
    else:
      self.workspace_node_list = self.wallpapoz_node.getElementsByTagName("file")

  ## class method -- get delay time that wallpaper list thread manipulate its index
  def delay(self):
    return self.wallpapoz_node.attributes["interval"].value

  ## class method -- whether the wallpaper list thread randomize its index
  def is_random(self):
    return self.wallpapoz_node.attributes["random"].value

  ## class method -- set delay time
  def set_delay(self, time):
    self.wallpapoz_node.setAttribute('interval', time)

  # set style attribute (fill screen, center, zoom, etc)
  def set_style(self, style):
    self.wallpapoz_node.setAttribute('style', style)

  # get style attribute (fill screen, center, zoom, etc)
  def style(self):
    try:
      return self.wallpapoz_node.attributes["style"].value
    except KeyError:
      # default is centered
      return "2"

  # get type attribute (desktop or workspace)
  def get_type(self):
    try:
      return self.wallpapoz_node.attributes["type"].value
    except KeyError:
      return "workspace"

  # get changing wallpaper when changing workspace
  def change_wallpaper_when_changing_workspace(self):
    try:
      if self.wallpapoz_node.attributes["type"].value == 'desktop':
        return False
      elif self.wallpapoz_node.attributes["type"].value == 'workspace':
        return True
    except KeyError:
      return True

  ## class method -- set random
  def set_random(self, randm):
    if randm:
      self.wallpapoz_node.setAttribute('random', "1" )
    else:
      self.wallpapoz_node.setAttribute('random', "0" )

  # fill list with wallpaper path file from xml file
  def fill_list(self):
    # container for wallpapers list
    worklist = []

    # if our configuration file is for workspace
    # I mean when user change workspace, the wallpaper change
    if self.wallpapoz_type == "workspace":

      # for every workspace we have in current desktop
      # so if we have 4 workspace, worklist will be like this:
      # [ [], [], [], [] ]
      for i in range(self.workspace_num):
        worklist.append( [] )

      # workspace number
      index = -1

      # remember this is our xml file
      # <workspace name="blabla">
      #   <file>bla.jpg</file>
      #   <file>buu.jpg</file>
      #   .....
      # </workspace>
      # <workspace name="gaga">
      #   <file>....

      # workspace_node hold <workspace name="blabla">
      for workspace_node in self.workspace_node_list:
        # file_node_list hold all <file> below <workspace...>
        file_node_list = workspace_node.getElementsByTagName('file')

        # our treeview example
        # workspace             wallpaper
        # >1                    nature
        #  >1                    sunset.jpg
        #  >2                    moon.png
        # >2                    sexy girls
        #  >1                    britney.jpg

        # the corresponding xml file will be like this:
        # <workspace name="nature">
        #   <file>sunset.jpg</file>
        #   <file>moon.png</file>
        # </workspace>
        # <workspace name="sexy girls">
        #   <file>britney.jpg</file>

        # index hold workspace number, that is number in the left of "nature" and "sexy girls"
        index = index + 1

        # if our workspace amount in desktop is lesser than workspace node in xml file
        # then no need to fill it again from xml file
        if len(worklist) <= index:
          break;

        # we put "name" attribute value to the first of every list inside worklist
        # <workspace name="stupid">, "name" attribute value will be "stupid"
        # so worklist will be like this in first iteration:
        # [ ["nature"], [], [], [] ]
        # example taken from above line "index = index + 1"
        worklist[index].append( workspace_node.attributes["name"].value )

        # node will hold single <file> below <workspace...>
        # we iterating all <file> below <workspace...>
        for node in file_node_list:
          # here we just interested in "file" not empty space
          # remember, this is our xml file
          # <workspace name="stupid">
          #   <file>bla.jpg</file>
          #   <file>buu.jpg</file>
          # .....
          # for first iteration, our node hold this value: emptyspace<file>bla.jpg</file>emptyspace
          # emptyspace is space between workspace node and file node
          # of course, if our xml file is like this:
          # <workspace name="stupid"><file>bla.jpg</file><file>buu.jpg</file>....
          # there is no emptyspace
          # but I don't like that xml file style
          if node.nodeName == "file":
            # put it in our worklist list
            worklist[index].append(node.firstChild.data)

          # remember our xml file, upthere
          # after iterating all of it, our worklist will be like this:
          # [ ["nature", "sunset.jpg", "moon.png"], [], [], [] ]
          # so worklist is designed to hold data like this:
          # [ [name_of_wallpapers_group1, wallpaper1, wallpaper2,...], [name_of_wallpapers_group2, wallpaper1,....],...]
          # remember, the first data of every list inside worklist list is name of wallpaper group like "nature",
          # "sexy girls". The second data and so on will hold wallpapers data

      # if our workspace in desktop is larger than workspace node amount in xml file
      # we must add the rest with "rename this" on wallpapers group name
      # and for every workspace of the rest give one wallpaper, that is current wallpaper
      # for example:
      # our workspace in desktop is 10
      # our workspace node in xml file is 8 ( maybe after creating the configuration file
      #                                       for the first time, we add the workspace )
      # our worklist is like this for the first time
      # [ ["nature", ...],...,["8th workspace",...], [], [] ]
      # after our worklist going through this iteration, it will be like this:
      # [ ["nature", ...],...,["8th workspace",...], ["rename this", current_wallpaper], ["rename this", current_wallpaper] ]
      index = index + 1
      if index < self.workspace_num:
        for i in range(self.workspace_num - index):
          worklist[index].append(_("rename this"))
          worklist[index].append(self.current_wallpaper)
          index += 1

    # if our configuration file is for desktop,
    # we don't change wallpaper when user change workspace
    else:
      # file_node hold <file>blabla.jpg</file>
      for file_node in self.workspace_node_list:
        # put it in our worklist list
        worklist.append(file_node.firstChild.data)

    return worklist

  # save the xml file
  def save(self):
    xml_file = open(self.config_file, "w")
    xml_file.write( self.xmldoc.toxml("utf-8") )

  # fill list with default value
  def default_fill_list(self, type):
    default_list = []
    if type == "treestore":
      for i in range(self.workspace_num):
        default_list.append([])
        default_list[i].append(_("rename this"))
        default_list[i].append(self.current_wallpaper)
    elif type == "liststore":
      default_list.append(self.current_wallpaper)
    return default_list

  ## class method -- create configuration file
  def create_configuration_file(self, wallpaperlist, tree_type):

    # use minidom implementation
    impl = minidom.getDOMImplementation()

    # this is the documenttype and document
    doctype = impl.createDocumentType("Wallpapoz", None, None)
    newdoc = impl.createDocument(None, "wallpapoz", doctype)

    # root element ( wallpapoz )
    top_element = newdoc.documentElement

    # pretty printing
    space_wallpapoz = newdoc.createTextNode("\n")
    top_element.appendChild(space_wallpapoz)
    # end of pretty printing

    # put attributes to wallpapoz element
    if self.xmldoc:
      top_element.setAttribute('random', self.is_random())
      top_element.setAttribute('interval', self.delay())
      top_element.setAttribute('style', self.style())
    else:
      top_element.setAttribute('random', '0')
      top_element.setAttribute('interval', '5')
      top_element.setAttribute('style', '2')

    # put type attributes to wallpapoz element and make the xml file
    if tree_type == "liststore":
      top_element.setAttribute('type', "desktop")

      # create file node based on wallpaper list
      for filepath in wallpaperlist:
        fileelement = newdoc.createElement("file")

        # pretty printing
        space_workspace_tab = newdoc.createTextNode("    ")
        top_element.appendChild(space_workspace_tab)
        # end of pretty printing

        # put wallpaper path to file node text
        wallpaper_path = newdoc.createTextNode(filepath)
        fileelement.appendChild(wallpaper_path)

        # add file node to top element
        top_element.appendChild(fileelement)

        # pretty printing
        space_file = newdoc.createTextNode("\n")
        top_element.appendChild(space_file)
        # end of pretty printing

    elif tree_type == "treestore":
      top_element.setAttribute('type', "workspace")

      index_of_wallpaperlist = 1

      # create workspace node based on wallpaper list
      for workspace in wallpaperlist:
        workspaceelement = newdoc.createElement("workspace")
        workspaceelement.setAttribute("no", str(index_of_wallpaperlist) )
        workspaceelement.setAttribute("name", workspace.pop(0) )

        # add the index
        index_of_wallpaperlist += 1

        # pretty printing
        space_workspace_tab = newdoc.createTextNode("    ")
        top_element.appendChild(space_workspace_tab)
        space_workspace = newdoc.createTextNode("\n")
        workspaceelement.appendChild(space_workspace)
        # end of pretty printing

        for file in workspace:

          # pretty printing
          space_file_tab = newdoc.createTextNode("        ")
          workspaceelement.appendChild(space_file_tab)
          # end of pretty printing

          fileelement = newdoc.createElement("file")
          path = newdoc.createTextNode(file)
          fileelement.appendChild(path)
          workspaceelement.appendChild(fileelement)

          # pretty printing
          space_file = newdoc.createTextNode("\n")
          workspaceelement.appendChild(space_file)
          # end of pretty printing

        # pretty printing
        space_file_tab = newdoc.createTextNode("    ")
        workspaceelement.appendChild(space_file_tab)
        # end of pretty printing

        top_element.appendChild(workspaceelement)

        # pretty printing
        space = newdoc.createTextNode("\n")
        top_element.appendChild(space)
        # end of pretty printing

    # create it
    xml_file = open(self.config_file, "w")
    xml_file.write( newdoc.toxml("utf-8") )
